crates/accelerate/src/synthesis/linear_phase/cx_cz_depth_lnn.rs

```diff
@@ -0,0 +1,284 @@
+                        && (k != j)
+                        && (order_comp[min(k, j)] < order_comp[i])
+                        && (phase_schedule[[min(k, j), max(k, j)]] % 4 != 0)
+                })
+                .collect();
+
+            for k in valid_indices {
+                let phase = phase_schedule[[min(k, j), max(k, j)]];
+                phase_schedule[[min(k, j), max(k, j)]] = 0;
+                // Step 1, apply phase to c_i, c_j, c_k
+                for l_s in [i, j, k] {
+                    phase_schedule[[l_s, l_s]] = (phase_schedule[[l_s, l_s]] + phase * 3) % 4;
+                }
+                // Step 2, apply phase to c_i+ c_j, c_i+c_k, c_j+c_k:
+                for (l1, l2) in [(i, j), (i, k), (j, k)] {
+                    let ls = min(l1, l2);
+                    let lb = max(l1, l2);
+                    phase_schedule[[ls, lb]] = (phase_schedule[[ls, lb]] + phase * 3) % 4;
+                }
+            }
+        });
+}
+
+/// Given
+///     Width of the circuit (int n)
+///     A CZ circuit, represented by the n*n phase schedule phase_schedule
+///     A CX circuit, represented by box-labels (seq) and whether the box is SWAP+ (swap_plus)
+///         *   This circuit corresponds to the CX tranformation that tranforms a matrix to
+///             a NW matrix (c.f. Prop.7.4, [1])
+///         *   SWAP+ is defined in section 3.A. of [2].
+///         *   As previously noted, the northwest diagonalization procedure of [1] consists
+///             of exactly n layers of boxes, each being either a SWAP or a SWAP+. That is,
+///             each northwest diagonalization circuit can be uniquely represented by which
+///             of its n(n-1)/2 boxes are SWAP+ and which are SWAP.
+/// Return a QuantumCircuit that computes the phase schedule S inside CX
+fn _apply_phase_to_nw_circuit(
+    n: usize,
+    phase_schedule: &Array2<usize>,
+    seq: &[(usize, usize)],
+    swap_plus: &HashSet<(usize, usize)>,
+) -> Vec<CircuitInstructions> {
+    let wires: Vec<_> = (0..n - 1)
+        .step_by(2)
+        .zip((1..n).step_by(2))
+        .chain((1..n - 1).step_by(2).zip((2..n).step_by(2)))
+        .collect();
+
+    let mut cir: Vec<CircuitInstructions> = Vec::new();
+    for (i, &(j, k)) in (0..seq.len()).rev().zip(seq.iter().rev()) {
+        let (w1, w2) = wires[i % (n - 1)];
+        if !swap_plus.contains(&(j, k)) {
+            cir.push(CircuitInstructions::CX(w1 as u32, w2 as u32));
+        }
+        cir.push(CircuitInstructions::CX(w2 as u32, w1 as u32));
+        match phase_schedule[[j, k]] % 4 {
+            0 => {}
+            1 => cir.push(CircuitInstructions::Sdg(w2 as u32)),
+            2 => cir.push(CircuitInstructions::Z(w2 as u32)),
+            3 => cir.push(CircuitInstructions::S(w2 as u32)),
+            _ => unreachable!(),
+        }
+        cir.push(CircuitInstructions::CX(w1 as u32, w2 as u32));
+    }
+    for i in 0..n {
+        match phase_schedule[[n - 1 - i, n - 1 - i]] {
```

Comment:
I see now I missed the `% 4` here and it wasn't caught by testing; I'll extend the tests to ensure this is caught.

Reply:
It seems this cannot have any real effect since the values in the phase schedule are always computed modulo 4 so the code is technically correct, although I'll add the `% 4` since it's better to be on the safe side.